src/hugr/rewrite.rs

```diff
@@ -0,0 +1,14 @@
+//! Replace operations on the HUGR.
+
+#[allow(clippy::module_inception)] // TODO: Rename?
+pub mod replace;
+use crate::Hugr;
+pub use replace::{OpenHugr, Replace, ReplaceError};
+
+/// An operation that can be applied to mutate a Hugr
+pub trait Rewrite<E> {
+    /// Mutate the specified Hugr, or fail with an error.
+    /// Implementations are strongly encouraged not to mutate the Hugr
+    /// if they return an Err.
```

Comment:
That's quite restrictive. Normally we'd use `?` to resurface any internal errors during the rewriting. Those may happen at any point, so this would require doing (potentially costly) dry-checks before running.

Reply:
Transactional semantics is it's own beast. Anything beyond 1. will incur a hit to performance, that we may perhaps prefer the user to have control over.

2. 3. 4. don't seem too practical without lots of extra overhead.
5. May be an option for region-focused rewrites, but not necessarily for every rewrite.
6. The pre-execution predicates are a thing in tket, and I think something we wanted to move away from.

I'd say the general rewrite interface shouldn't require this. I propose:

7. Leave these assurances to each Replace operation. Add `Transactional<>` wrappers or perhaps specialized versions (e.g. `struct TransactionalSimpleReplace(SimpleReplace)`) so the users that actually want to pay the cost can choose to.

@cqc-alec and @ss2165 may have opinions, but we could move this to a discussion too.

Reply:
Hmmm, perhaps I should state that `apply` may `panic` if-and-only-if the Hugr doesn't validate()?

Reply:
Ok, I think I've enacted as discussed